File: stack/valid-parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        braces = {
            "(": ")",
            "[": "]",
            "{": "}",
        }
        stack = []
        for brace in s:
            if brace in braces:
                stack.append(braces[brace])
            else:
                if len(stack) == 0 or stack.pop() != brace:
                    return False
        return len(stack) == 0

File: string/valid-parentheses.rb
# @param {String} s
# @return {Boolean}
def is_valid(s)
    mapping = {
        "(" => ")",
        "[" => "]",
        "{" => "}",
    }
    stack = []
    s.each_char do |brace|
        if mapping.include? brace
            stack << mapping[brace]
        else
            return false if stack.empty? || stack.pop != brace
        end
    end
    stack.length.zero?
end